hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/CommonUtil.java

```diff
@@ -0,0 +1,410 @@
+     * The timeout can be specified by the parameter.
+     *
+     * @param attempt      an attempt
+     * @param finalState   the attempt state waited
+     * @param timeoutMsecs the length of timeout in milliseconds
+     * @throws InterruptedException if interrupted while waiting for the state transition
+     */
+    public static void waitForState(RMAppAttempt attempt,
+                                    RMAppAttemptState finalState, int timeoutMsecs)
+            throws InterruptedException {
+        int timeWaiting = 0;
+        while (finalState != attempt.getAppAttemptState()) {
+            if (timeWaiting >= timeoutMsecs) {
+                break;
+            }
+
+            LOG.info("AppAttempt : " + attempt.getAppAttemptId() + " State is : " +
+                    attempt.getAppAttemptState() + " Waiting for state : " + finalState);
+            Thread.sleep(WAIT_MS_PER_LOOP);
+            timeWaiting += WAIT_MS_PER_LOOP;
+        }
+
+        LOG.info("Attempt State is : " + attempt.getAppAttemptState());
+        Assert.assertEquals("Attempt state is not correct (timeout).", finalState,
+                attempt.getState());
+    }
+
+    public static void waitForContainerToComplete(MockRM rm, RMAppAttempt attempt,
+                                           NMContainerStatus completedContainer) throws InterruptedException {
+        rm.drainEventsImplicitly();
+        int timeWaiting = 0;
+        while (timeWaiting < TIMEOUT_MS_FOR_CONTAINER_AND_NODE) {
+            List<ContainerStatus> containers = attempt.getJustFinishedContainers();
+            LOG.info("Received completed containers " + containers);
+            for (ContainerStatus container : containers) {
+                if (container.getContainerId().equals(
+                        completedContainer.getContainerId())) {
+                    return;
+                }
+            }
+            Thread.sleep(WAIT_MS_PER_LOOP);
+            timeWaiting += WAIT_MS_PER_LOOP;
+        }
+    }
+
+    public static MockAM waitForNewAMToLaunchAndRegister(MockRM rm, ApplicationId appId, int attemptSize,
+                                                  MockNM nm) throws Exception {
+        RMApp app = rm.getRMContext().getRMApps().get(appId);
+        Assert.assertNotNull(app);
+        int timeWaiting = 0;
+        while (app.getAppAttempts().size() != attemptSize) {
+            if (timeWaiting >= TIMEOUT_MS_FOR_ATTEMPT) {
+                break;
+            }
+            LOG.info("Application " + appId
+                    + " is waiting for AM to restart. Current has "
+                    + app.getAppAttempts().size() + " attempts.");
+            Thread.sleep(WAIT_MS_PER_LOOP);
+            timeWaiting += WAIT_MS_PER_LOOP;
+        }
+        return rm.launchAndRegisterAM(app, rm, nm);
+    }
+
+    /**
+     * Wait until a container has reached a specified state.
+     * The timeout is 10 seconds.
```

Comment:
@K0K0V0K The timeout is specified by the parameter.They already mentioned 10, maybe it was calculated or general timeout.

Reply:
So i mean in 2017 the value of the **TIMEOUT_MS_FOR_CONTAINER_AND_NODE** was increased from 10 to 20 with the YARN-5876 ticket, but the author forget to update this javadoc.
To prevent these scenarios we can use the @link annotation instead of hardcoded values.